### PR TITLE
MM-12354 Add ability to remove profile image

### PR DESCRIPTION
**Pull request by bradjcoughlin**
#### Summary
Add ability to remove profile image

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-12354

#### Checklist
- [ ] Added or updated unit tests (required for all new features)
- [x] Has UI changes
- [x] Includes text changes and localization file updates

#### Device Information
This PR was tested on: iPhone 6 Simulator, iOS 12.1

#### Screenshots
screen shot 2018-11-17 at 12 05 42 pm
screen shot 2018-11-17 at 12 06 08 pm

### CODE REVIEW

File: app/actions/views/edit_profile.js [hunk 1]
```diff
@@ -25,7 +25,21 @@ export function setProfileImageUri(imageUri = '') {
     };
 }
 
+export function removeProfileImage(user, success, error) {
```
Comment (enahum) [inline app/actions/views/edit_profile.js]: Why using callbacks here?

File: app/components/attachment_button.js [hunk 2]
```diff
@@ -328,6 +330,20 @@ export default class AttachmentButton extends PureComponent {
...
 
+    removeProfileImage = () => {
+        this.props.removeProfileImage();
+        this.props.navigator.dismissModal({
+            animationType: 'none',
+        });
+    };
+
+    isCustomProfileImage = () => {
+        if (this.props.profileImageUri.includes('?')) {
```
Comment (enahum) [inline app/components/attachment_button.js]: Why is this needed?

File: app/components/attachment_button.js [hunk 3]
```diff
@@ -382,6 +398,25 @@ export default class AttachmentButton extends PureComponent {
             icon: 'file',
         });
 
+        // Show option to remove image in edit profile screen
+        // default image will not include query string with timestamp
+        if (this.props.removeProfileImage && this.isCustomProfileImage()) {
+            options.items.push({
+                action: () => this.removeProfileImage(),
```
Comment (enahum) [inline app/components/attachment_button.js]: Can you just do action: this.removeProfileImage without the arrow function?

File: app/screens/edit_profile/edit_profile.js [hunk 4]
```diff
@@ -87,16 +88,29 @@ export default class EditProfile extends PureComponent {
...
             position,
             username,
+            profileImageUri,
         };
     }
 
+    getProfileImageUri = async (id, lastPictureUpdate) => {
+        let uri;
+        try {
+            uri = await Client4.getProfilePictureUrl(id, lastPictureUpdate);
```
Comment (enahum) [inline app/screens/edit_profile/edit_profile.js]: Why are we using the client here directly? In any case this should be an action

File: app/screens/edit_profile/edit_profile.js [hunk 5]
```diff
@@ -221,6 +235,13 @@ export default class EditProfile extends PureComponent {
         return RNFetchBlob.config(options).fetch('POST', `${Client4.getUserRoute(currentUser.id)}/image`, headers, [fileInfo]);
     };
 
+    removeProfileImage = async () => {
+        const {actions} = this.props;
+        const {currentUser} = this.props;
+        this.emitCanUpdateAccount(true);
```
Comment (enahum) [inline app/screens/edit_profile/edit_profile.js]: Is it really needed? You are already removing the image from the server, aren’t you?

File: app/screens/edit_profile/edit_profile.js [hunk 6]
```diff
@@ -87,16 +88,29 @@ export default class EditProfile extends PureComponent {
         props.navigator.setOnNavigatorEvent(this.onNavigatorEvent);
         props.navigator.setButtons(buttons);
 
+        const profileImageUri = '';
```
Comment (enahum) [inline app/screens/edit_profile/edit_profile.js]: What does this do?

File: app/screens/edit_profile/edit_profile.js [hunk 7]
```diff
@@ -436,6 +457,11 @@ export default class EditProfile extends PureComponent {
         this.scrollView = ref;
     };
 
+    componentDidMount() {
+        const {id, last_picture_update: lastPictureUpdate} = this.props.currentUser;
+        this.getProfileImageUri(id, lastPictureUpdate);
```
Comment (enahum) [inline app/screens/edit_profile/edit_profile.js]: What is the purpose of this? Don’t you already have the image uri?

File: app/components/attachment_button.js (hunk 3 above)
Comment (bradjcoughlin) [reply app/components/attachment_button.js]: sure, originally i thought i would need a param there.

File: app/components/attachment_button.js (hunk 2 above)
Comment (bradjcoughlin) [reply app/components/attachment_button.js]: Attempting to identify the default image vs custom image URL. It looks like the only difference between the image URL returned is the query string with the last modified timestamp. There would be no need to show the "Remove Image" option if the user profile has the default image set.

File: app/screens/edit_profile/edit_profile.js (hunk 7 above)
Comment (bradjcoughlin) [reply app/screens/edit_profile/edit_profile.js]: I don't have the image URI from what i can see in the currentUser object.

File: app/screens/edit_profile/edit_profile.js (hunk 4 above)
Comment (bradjcoughlin) [reply app/screens/edit_profile/edit_profile.js]: that's what i wondered. i can refactor this. the main reason is to get the profile image URL because I don't have it in the currentUser object from what i can see. as i mentioned on the other comment, there would be no need to show the "Remove Image" option if the user profile has the default image set.

File: app/screens/edit_profile/edit_profile.js (hunk 5 above)
Comment (bradjcoughlin) [reply app/screens/edit_profile/edit_profile.js]: This is calling the action that then calls 'setDefaultProfileImage'. Is there another method to remove the image? It worked how I expected it would.

File: app/screens/edit_profile/edit_profile.js (hunk 4 above)
Comment (bradjcoughlin) [reply app/screens/edit_profile/edit_profile.js]: shouldn't 'uploadProfileImage' be an action as well?

File: app/screens/edit_profile/edit_profile.js (hunk 6 above)
Comment (bradjcoughlin) [reply app/screens/edit_profile/edit_profile.js]: looks like nothing. removing.

File: app/components/attachment_button.js (hunk 2 above)
Comment (enahum) [reply app/components/attachment_button.js]: yeah I thought it would have the timestamp set to 0 but you are right here

File: app/components/attachment_button.js [hunk 8]
```diff
@@ -328,6 +330,20 @@ export default class AttachmentButton extends PureComponent {
         }, 100);
     };
 
+    removeProfileImage = () => {
+        this.props.removeProfileImage();
```
Comment (enahum) [inline app/components/attachment_button.js]: the prop for removeProfile image can be null and you are not validating that here, potential crash

File: app/screens/edit_profile/edit_profile.js (hunk 4 above)
Comment (enahum) [reply app/screens/edit_profile/edit_profile.js]: actually seems that we do not have a selector for this, so I guess this is fine, in any case we could improve this later on, not needed for this PR and perhaps take a look at https://github.com/mattermost/mattermost-mobile/blob/master/app/components/profile_picture/profile_picture.js#L60 as you may use a local cache for the image instead of performing another request

File: app/screens/edit_profile/edit_profile.js (hunk 5 above)
Comment (enahum) [reply app/screens/edit_profile/edit_profile.js]: Not another method.

I don't quite follow, will you remove the image once the user selects the option or will you remove the image when saving? cause this is to enable the save option as far as I remember.

I would expect that you remove the image then when tapping on save it actually removes it from the server, if you cancel this the current image should remain.

Thoughts?

File: app/screens/edit_profile/edit_profile.js (hunk 7 above)
Comment (enahum) [reply app/screens/edit_profile/edit_profile.js]: Yeah this just returns the path, I thought it was making a request, sorry for the confusion

File: app/components/attachment_button.js (hunk 2 above)
Comment (enahum) [inline app/components/attachment_button.js]: this one will not cause a crash as the one above as you are checking for the prop before calling this function so you are good here

File: app/screens/edit_profile/edit_profile.js (hunk 4 above)
Comment (hmhealey) [reply app/screens/edit_profile/edit_profile.js]: This isn't actually making a client call. It's just a helper method to construct the URL that happens to be on the Client4 instance. It could be moved into the utils or something, but we've kept all the URL construction logic in the client for a long time, so I wouldn't worry about changing it now

File: app/components/attachment_button.js [hunk 9]
```diff
@@ -35,6 +35,8 @@ export default class AttachmentButton extends PureComponent {
         onShowFileSizeWarning: PropTypes.func,
         theme: PropTypes.object.isRequired,
         uploadFiles: PropTypes.func.isRequired,
+        removeProfileImage: PropTypes.func,
```
Comment (hmhealey) [inline app/components/attachment_button.js]: We also use this component in other places, so we should try to keep this behaviour separate from the component itself. What you could do is instead of adding props for `removeProfileImage` and `profileImageUri`, you could instead add a prop like `getExtraActions` so that the `EditProfile` screen can tell the `AttachmentButton` about the extra option to render.

Also, this part isn't really needed, but you could even take it one step further and make a separate component that wraps `AttachmentButton` to contain that logic and make the `EditProfile` screen a bit simpler. It could look like
```javascript
class ProfilePictureButton extends React.PureComponent {
    static propTypes = {
        currentUser: PropTypes.object.isRequired,
    };

    getExtraOptions = () => {
    	if (!this.isCustomProfileImage()) {
    		return null;
    	}

    	return [{
	    	action: this.removeProfileImage,
	    	text: 'Remove Image'
    	}];
    }

    render() {
	    const {children, ...props} = this.props;

	    // Avoid passing unneeded props
	    Reflect.deleteProperty(props, 'currentUser');

        return (
            <AttachmentButton
            	{...props}
            	getExtraOptions={this.getExtraOptions}
            >
            	{children}
            </AttachmentButton>
        );
    }
}
```

File: app/screens/edit_profile/edit_profile.js [hunk 10]
```diff
@@ -87,16 +88,29 @@ export default class EditProfile extends PureComponent {
...
             firstName,
             lastName,
             nickname,
             position,
             username,
+            profileImageUri,
         };
     }
 
+    getProfileImageUri = async (id, lastPictureUpdate) => {
```
Comment (hmhealey) [inline app/screens/edit_profile/edit_profile.js]: The async/await and the try/catch aren't actually needed here since `Client4.getProfilePictureUrl` is synchronous

File: app/components/attachment_button.js (hunk 9 above)
Comment (bradjcoughlin) [reply app/components/attachment_button.js]: I also wondered about that, as it is a global component. That looks like a good approach to keep it clean.

File: app/actions/views/edit_profile.js (hunk 1 above)
Comment (bradjcoughlin) [reply app/actions/views/edit_profile.js]: I may not fully grasp the underlying redux action, but the 'updateUser' action on edit_profile.js line 8 is using the callbacks as well.

File: app/screens/edit_profile/edit_profile.js (hunk 5 above)
Comment (bradjcoughlin) [reply app/screens/edit_profile/edit_profile.js]: On second thought, I understand what you are asking now @enahum . It doesn't seem like I need to enable the Save button in the UI, which is what this event is triggering. So, you're right, I can take this out.

**Comment by enahum** [comment]
@bradjcoughlin any progress on this? If we don't merge this by tomorrow it will need to wait until the next feature release which is in January

**Comment by hmhealey** [comment]
Not at all a problem. It's totally fine to just release it a bit later

File: app/components/attachment_button.js (hunk 3 above)
Comment (enahum) [reply app/components/attachment_button.js]: not really you should do

```js
options.items.push({
    action: this.removeProfileImage,
   ...
});
```

Comment (bradjcoughlin) [reply app/components/attachment_button.js]: Makes sense - since this is just a reference and the binding happens on the lower level component.

**Comment by bradjcoughlin** [comment]
I'm going to close this and submit and new PR if that's ok.
